File: C++/ScientificCalculator/src/Tokenizer.cpp
#include "Tokenizer.h"

#include <cctype>
#include <stdexcept>


vector<string> tokenize(
    const string& expression
)
{

    vector<string> tokens;


    for(size_t i = 0; i < expression.length();)
    {

        char c = expression[i];


        // Ignore spaces

        if(isspace(c))
        {
            i++;
            continue;
        }



        // Numbers

        if(
            isdigit(c) ||
            c == '.'
        )
        {

            string number;


            while(
                i < expression.length() &&
                (
                    isdigit(expression[i]) ||
                    expression[i] == '.'
                )
            )
            {

                number += expression[i];

                i++;

            }


            tokens.push_back(number);

            continue;
        }



        // Names:
        // sqrt, sin, root, pi, e

        if(isalpha(c))
        {

            string name;


            while(
                i < expression.length() &&
                isalpha(expression[i])
            )
            {

                name += expression[i];

                i++;

            }


            tokens.push_back(name);

            continue;
        }



        // Operators and brackets

        if(
            c == '+' ||
            c == '-' ||
            c == '*' ||
            c == '/' ||
            c == '%' ||
            c == '^' ||
            c == '(' ||
            c == ')' ||
            c == ','
        )
        {

            tokens.push_back(
                string(1,c)
            );

            i++;

            continue;
        }



        throw runtime_error(
            string("Unknown character: ") + c
        );

    }


    return tokens;
}

File: Java/ScientificCalculator/src/parser/Tokenizer.java
package parser;

import java.util.ArrayList;
import java.util.List;

public class Tokenizer {

    // ==========================
    // Tokenize Expression
    // ==========================

    public static List<String> tokenize(String expression) {

        List<String> tokens = new ArrayList<>();

        StringBuilder current = new StringBuilder();

        for (int i = 0; i < expression.length(); i++) {

            char c = expression.charAt(i);

            // ==========================
            // Numbers / Names
            // ==========================

            if (Character.isDigit(c)
                    || c == '.'
                    || Character.isLetter(c)) {

                current.append(c);
                continue;
            }

            // Finish current token

            if (current.length() > 0) {

                tokens.add(current.toString());
                current.setLength(0);

            }

            // ==========================
            // Spaces
            // ==========================

            if (Character.isWhitespace(c)) {
                continue;
            }

            // ==========================
            // Operators / Brackets / Comma
            // ==========================

            if (
                    c == '+'
                    || c == '-'
                    || c == '*'
                    || c == '/'
                    || c == '^'
                    || c == '('
                    || c == ')'
                    || c == ','
            ) {

                tokens.add(String.valueOf(c));

            } else {

                throw new RuntimeException(
                        "Unknown character: " + c
                );

            }

        }

        if (current.length() > 0) {

            tokens.add(current.toString());

        }

        return tokens;

    }

}
